File: macos/Sources/Services/Protocol.swift
import CommonCrypto
import Foundation
import os.log

private let log = OSLog(subsystem: "com.buzzel", category: "Protocol")

// MARK: - Signal Identifiers

enum Signal {
    static let command: UInt8 = 0x00
    static let pairRequest: UInt8 = 0x01
    static let pairResponse: UInt8 = 0x02
    static let ready: UInt8 = 0x03
    static let ping: UInt8 = 0x04
    static let pong: UInt8 = 0x05
    static let acknowledgment: UInt8 = 0x06
    static let goodbye: UInt8 = 0x07
    static let unpair: UInt8 = 0x08
}

// MARK: - Bluetooth Low Energy UUIDs

enum BluetoothServiceUUIDs {
    static let service = "0000BF01-0000-1000-8000-00805F9B34FB"
    static let dataCharacteristic = "0000BF02-0000-1000-8000-00805F9B34FB"
}

// MARK: - Protocol

enum BuzzelProtocol {
    static let tcpPort: UInt16 = 48155
    static let frameHeaderSize = 2
    static let commandHeaderSize = 3  // sequenceNumber(2) + commandIdentifier(1)
    static let attributeProtocolOverhead = 3
    static let tagLengthValueMaximumValueSize = 255  // 1-byte length field
    static let wifiMaximumFrameSize = 4096

    static func maximumPayloadSize(_ maximumFrameSize: Int) -> Int {
        maximumFrameSize - frameHeaderSize
    }

    static func maximumCommandDataSize(_ maximumFrameSize: Int) -> Int {
        maximumPayloadSize(maximumFrameSize) - 1 - commandHeaderSize
    }

    static func maximumTagLengthValueDataSize(_ maximumFrameSize: Int) -> Int {
        min(maximumCommandDataSize(maximumFrameSize) - 2, tagLengthValueMaximumValueSize)
    }

    static func maximumFrameSizeForMaximumTransmissionUnit(_ maximumTransmissionUnit: Int) -> Int {
        maximumTransmissionUnit - attributeProtocolOverhead
    }

    // MARK: QR Code

    static let qrCodePayloadSize = 24
    private static let qrCodeMagicNumber: UInt16 = 0xBC1B

    struct QRCodePayload {
        let seed: Data
        let host: String
        let preferredTransport: String
    }

    static func generateQRCodePayload(seed: Data, host: String, preferredTransport: String) -> Data {
        var buffer = Data(capacity: qrCodePayloadSize)
        var magicNumber = qrCodeMagicNumber.bigEndian
        buffer.append(Data(bytes: &magicNumber, count: 2))
        buffer.append(seed)
        let hostParts = host.split(separator: ".").compactMap { UInt8($0) }
        if hostParts.count == 4 {
            buffer.append(contentsOf: hostParts)
        } else {
            buffer.append(contentsOf: [0, 0, 0, 0])
        }
        buffer.append(preferredTransport == "ble" ? 0x01 : 0x00)
        buffer.append(0x00)  // reserved
        return buffer
    }

    static func deriveSessionIdentifier(_ seed: Data) -> String {
        let hash = computeSHA256(seed)
        let uuid = NSUUID(uuidBytes: [UInt8](hash.prefix(16)))
        return uuid.uuidString
    }

    static func derivePairingCode(_ seed: Data) -> String {
        var input = seed
        input.append("code".data(using: .utf8)!)
        let hash = computeSHA256(input)
        let byte0 = UInt32(hash[0]) << 24
        let byte1 = UInt32(hash[1]) << 16
        let byte2 = UInt32(hash[2]) << 8
        let byte3 = UInt32(hash[3])
        let numericValue = byte0 | byte1 | byte2 | byte3
        let code = numericValue % 1_000_000
        return String(format: "%06d", code)
    }

    private static func computeSHA256(_ data: Data) -> Data {
        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        _ = data.withUnsafeBytes { CC_SHA256($0.baseAddress, CC_LONG(data.count), &hash) }
        return Data(hash)
    }

    // MARK: Signals

    static func createPing() -> Data {
        Data([Signal.ping])
    }

    static func createPong() -> Data {
        Data([Signal.pong])
    }

    static func createReady() -> Data {
        Data([Signal.ready])
    }

    static func createGoodbye() -> Data {
        Data([Signal.goodbye])
    }

    static func createUnpair() -> Data {
        Data([Signal.unpair])
    }

    static func createPairRequest(code: String) -> Data {
        var buffer = Data(capacity: 7)
        buffer.append(Signal.pairRequest)
        let codeBytes = Array(code.utf8.prefix(6))
        buffer.append(contentsOf: codeBytes)
        let paddingNeeded = 7 - buffer.count
        if paddingNeeded > 0 { buffer.append(contentsOf: [UInt8](repeating: 0x30, count: paddingNeeded)) }
        return buffer
    }

    static func createPairResponse(accepted: Bool, reason: UInt8 = 0x00) -> Data {
        Data([Signal.pairResponse, accepted ? 0x01 : 0x00, reason])
    }

    static func createAcknowledgment(sequenceNumber: UInt16) -> Data {
        var buffer = Data(capacity: 3)
        buffer.append(Signal.acknowledgment)
        buffer.append(UInt8(sequenceNumber >> 8))
        buffer.append(UInt8(sequenceNumber & 0xFF))
        return buffer
    }

    static func parseSignalIdentifier(_ payload: Data) -> UInt8? {
        guard !payload.isEmpty else { return nil }
        return payload[0]
    }

    static func parsePairRequestCode(_ payload: Data) -> String? {
        guard payload.count >= 7, payload[0] == Signal.pairRequest else { return nil }
        return String(data: payload.subdata(in: 1..<7), encoding: .ascii)
    }

    static func parsePairResponse(_ payload: Data) -> (accepted: Bool, reason: UInt8)? {
        guard payload.count >= 3, payload[0] == Signal.pairResponse else { return nil }
        return (payload[1] == 0x01, payload[2])
    }

    static func parseAcknowledgmentSequenceNumber(_ payload: Data) -> UInt16? {
        guard payload.count >= 3, payload[0] == Signal.acknowledgment else { return nil }
        let highByte = UInt16(payload[1]) << 8
        let lowByte = UInt16(payload[2])
        return highByte | lowByte
    }

    // MARK: Commands

    struct Command {
        let commandIdentifier: UInt8
        let sequenceNumber: UInt16
        let data: Data
    }

    static func createCommand(
        commandIdentifier: UInt8, sequenceNumber: UInt16, tagLengthValueData: Data = Data(),
        maximumTagLengthValueDataSize: Int = tagLengthValueMaximumValueSize
    ) -> Data {
        let dataLength = min(tagLengthValueData.count, maximumTagLengthValueDataSize)
        var buffer = Data(capacity: 1 + commandHeaderSize + dataLength)
        buffer.append(Signal.command)
        buffer.append(UInt8(sequenceNumber >> 8))
        buffer.append(UInt8(sequenceNumber & 0xFF))
        buffer.append(commandIdentifier)
        if dataLength > 0 { buffer.append(tagLengthValueData.prefix(dataLength)) }
        return buffer
    }

    static func parseCommand(_ payload: Data) -> Command? {
        guard payload.count >= 4, payload[0] == Signal.command else { return nil }
        let sequenceHighByte = UInt16(payload[1]) << 8
        let sequenceLowByte = UInt16(payload[2])
        let sequenceNumber = sequenceHighByte | sequenceLowByte
        let commandIdentifier = payload[3]
        let data = payload.count > 4 ? payload.subdata(in: 4..<payload.count) : Data()
        return Command(
            commandIdentifier: commandIdentifier, sequenceNumber: sequenceNumber, data: data
        )
    }

    // MARK: QR Code Parsing

    static func parseQRCodePayload(_ rawData: Data) -> QRCodePayload? {
        guard rawData.count == qrCodePayloadSize else { return nil }
        let magicHighByte = UInt16(rawData[0]) << 8
        let magicLowByte = UInt16(rawData[1])
        let magicNumber = magicHighByte | magicLowByte
        guard magicNumber == qrCodeMagicNumber else { return nil }
        let seed = rawData.subdata(in: 2..<18)
        let host = "\(rawData[18]).\(rawData[19]).\(rawData[20]).\(rawData[21])"
        let preferredTransport = rawData[22] == 0x01 ? "ble" : "wifi"
        return QRCodePayload(seed: seed, host: host, preferredTransport: preferredTransport)
    }

    // MARK: Tag-Length-Value Encoding

    struct TagLengthValueField {
        let tag: UInt8
        let value: Data
    }

    static func encodeTagLengthValue(tag: UInt8, value: Data) -> Data {
        let length = min(value.count, tagLengthValueMaximumValueSize)
        var buffer = Data(capacity: 2 + length)
        buffer.append(tag)
        buffer.append(UInt8(length))
        buffer.append(value.prefix(length))
        return buffer
    }

    static func encodeTagLengthValueString(tag: UInt8, value: String) -> Data {
        encodeTagLengthValue(tag: tag, value: value.data(using: .utf8) ?? Data())
    }

    static func encodeTagLengthValueInteger(tag: UInt8, value: Int) -> Data {
        var bigEndianValue = UInt32(bitPattern: Int32(value)).bigEndian
        let bytes = Data(bytes: &bigEndianValue, count: 4)
        return encodeTagLengthValue(tag: tag, value: bytes)
    }

    static func encodeTagLengthValueByte(tag: UInt8, value: UInt8) -> Data {
        encodeTagLengthValue(tag: tag, value: Data([value]))
    }

    static func decodeTagLengthValue(_ data: Data) -> [TagLengthValueField] {
        var fields: [TagLengthValueField] = []
        var offset = 0
        while offset + 2 <= data.count {
            let tag = data[offset]
            let length = Int(data[offset + 1])
            offset += 2
            if offset + length > data.count { break }
            let value = data.subdata(in: offset..<(offset + length))
            fields.append(TagLengthValueField(tag: tag, value: value))
            offset += length
        }
        return fields
    }

    static func getTagLengthValueString(
        _ fields: [TagLengthValueField], tag: UInt8
    ) -> String? {
        guard let field = fields.first(where: { $0.tag == tag }) else { return nil }
        return String(data: field.value, encoding: .utf8)
    }

    static func getTagLengthValueInteger(
        _ fields: [TagLengthValueField], tag: UInt8
    ) -> Int? {
        guard let field = fields.first(where: { $0.tag == tag }), field.value.count >= 4 else {
            return nil
        }
        let byte0 = UInt32(field.value[0]) << 24
        let byte1 = UInt32(field.value[1]) << 16
        let byte2 = UInt32(field.value[2]) << 8
        let byte3 = UInt32(field.value[3])
        let rawValue = byte0 | byte1 | byte2 | byte3
        return Int(Int32(bitPattern: rawValue))
    }

    static func getTagLengthValueByte(
        _ fields: [TagLengthValueField], tag: UInt8
    ) -> UInt8? {
        guard let field = fields.first(where: { $0.tag == tag }), !field.value.isEmpty else {
            return nil
        }
        return field.value[0]
    }
}

File: android/app/src/main/java/com/buzzel/protocol/Protocol.kt
package com.buzzel.protocol

import com.buzzel.debug.FileLogger
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest
import java.util.UUID

object Signal {
    const val COMMAND: Byte = 0x00
    const val PAIR_REQUEST: Byte = 0x01
    const val PAIR_RESPONSE: Byte = 0x02
    const val READY: Byte = 0x03
    const val PING: Byte = 0x04
    const val PONG: Byte = 0x05
    const val ACKNOWLEDGMENT: Byte = 0x06
    const val GOODBYE: Byte = 0x07
    const val UNPAIR: Byte = 0x08
}

object BluetoothServiceUUIDs {
    val SERVICE: UUID = UUID.fromString("0000bf01-0000-1000-8000-00805f9b34fb")
    val DATA_CHARACTERISTIC: UUID = UUID.fromString("0000bf02-0000-1000-8000-00805f9b34fb")
    val CCCD: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
}

object Protocol {
    private const val TAG = "Protocol"

    const val TCP_PORT = 48155
    const val FRAME_HEADER_SIZE = 2
    const val COMMAND_HEADER_SIZE = 3 // sequenceNumber(2) + commandIdentifier(1)
    const val ATTRIBUTE_PROTOCOL_OVERHEAD = 3
    const val TAG_LENGTH_VALUE_MAXIMUM_VALUE_SIZE = 255 // 1-byte length field
    const val WIFI_MAXIMUM_FRAME_SIZE = 4096

    fun maximumPayloadSize(maximumFrameSize: Int): Int = maximumFrameSize - FRAME_HEADER_SIZE

    fun maximumCommandDataSize(maximumFrameSize: Int): Int =
        maximumPayloadSize(maximumFrameSize) - 1 - COMMAND_HEADER_SIZE

    fun maximumTagLengthValueDataSize(maximumFrameSize: Int): Int =
        minOf(maximumCommandDataSize(maximumFrameSize) - 2, TAG_LENGTH_VALUE_MAXIMUM_VALUE_SIZE)

    fun maximumFrameSizeForMaximumTransmissionUnit(maximumTransmissionUnit: Int): Int =
        maximumTransmissionUnit - ATTRIBUTE_PROTOCOL_OVERHEAD

    private const val QR_CODE_PAYLOAD_SIZE = 24
    private const val QR_CODE_MAGIC_NUMBER = 0xBC1B

    data class QRCodePayload(
        val seed: ByteArray,
        val host: String,
        val preferredTransport: String,
    )

    fun parseQRCodePayload(raw: ByteArray): QRCodePayload? {
        if (raw.size != QR_CODE_PAYLOAD_SIZE) {
            FileLogger.w(TAG, "QR: invalid size ${raw.size}")
            return null
        }
        val buffer = ByteBuffer.wrap(raw).order(ByteOrder.BIG_ENDIAN)
        val magicNumber = buffer.short.toInt() and 0xFFFF
        if (magicNumber != QR_CODE_MAGIC_NUMBER) {
            FileLogger.w(TAG, "QR: invalid magic 0x${magicNumber.toString(16)}")
            return null
        }
        val seed = ByteArray(16)
        buffer.get(seed)
        val hostBytes = ByteArray(4)
        buffer.get(hostBytes)
        val host = hostBytes.joinToString(".") { (it.toInt() and 0xFF).toString() }
        val preferredTransport = if (buffer.get().toInt() == 0x01) "ble" else "wifi"
        return QRCodePayload(seed, host, preferredTransport)
    }

    fun deriveSessionIdentifier(seed: ByteArray): String {
        val hash = computeSHA256(seed)
        val buffer = ByteBuffer.wrap(hash, 0, 16)
        val mostSignificantBits = buffer.long
        val leastSignificantBits = buffer.long
        return UUID(mostSignificantBits, leastSignificantBits).toString().uppercase()
    }

    fun derivePairingCode(seed: ByteArray): String {
        val input = seed + "code".toByteArray(Charsets.UTF_8)
        val hash = computeSHA256(input)
        val numericValue =
            ByteBuffer
                .wrap(hash, 0, 4)
                .order(ByteOrder.BIG_ENDIAN)
                .int
                .toLong() and 0xFFFFFFFFL
        return (numericValue % 1_000_000).toString().padStart(6, '0')
    }

    private fun computeSHA256(data: ByteArray): ByteArray = MessageDigest.getInstance("SHA-256").digest(data)

    fun createPing(): ByteArray = byteArrayOf(Signal.PING)

    fun createPong(): ByteArray = byteArrayOf(Signal.PONG)

    fun createReady(): ByteArray = byteArrayOf(Signal.READY)

    fun createGoodbye(): ByteArray = byteArrayOf(Signal.GOODBYE)

    fun createUnpair(): ByteArray = byteArrayOf(Signal.UNPAIR)

    fun createPairRequest(code: String): ByteArray {
        val codeBytes = code.toByteArray(Charsets.US_ASCII)
        val payload = ByteArray(1 + 6)
        payload[0] = Signal.PAIR_REQUEST
        System.arraycopy(codeBytes, 0, payload, 1, minOf(codeBytes.size, 6))
        return payload
    }

    fun createPairResponse(
        accepted: Boolean,
        reason: Byte = 0x00,
    ): ByteArray =
        byteArrayOf(
            Signal.PAIR_RESPONSE,
            if (accepted) 0x01 else 0x00,
            reason,
        )

    fun createAcknowledgment(sequenceNumber: Int): ByteArray {
        val buffer = ByteBuffer.allocate(3).order(ByteOrder.BIG_ENDIAN)
        buffer.put(Signal.ACKNOWLEDGMENT)
        buffer.putShort(sequenceNumber.toShort())
        return buffer.array()
    }

    fun parseSignalIdentifier(payload: ByteArray): Byte {
        if (payload.isEmpty()) return -1
        return payload[0]
    }

    fun parsePairRequestCode(payload: ByteArray): String? {
        if (payload.size < 7 || payload[0] != Signal.PAIR_REQUEST) return null
        return String(payload, 1, 6, Charsets.US_ASCII)
    }

    fun parsePairResponse(payload: ByteArray): Pair<Boolean, Byte>? {
        if (payload.size < 3 || payload[0] != Signal.PAIR_RESPONSE) return null
        val accepted = payload[1] == 0x01.toByte()
        val reason = payload[2]
        return Pair(accepted, reason)
    }

    fun parseAcknowledgmentSequenceNumber(payload: ByteArray): Int? {
        if (payload.size < 3 || payload[0] != Signal.ACKNOWLEDGMENT) return null
        return ByteBuffer
            .wrap(payload, 1, 2)
            .order(ByteOrder.BIG_ENDIAN)
            .short
            .toInt() and 0xFFFF
    }

    fun createCommand(
        commandIdentifier: Byte,
        sequenceNumber: Int,
        tagLengthValueData: ByteArray = ByteArray(0),
        maximumTagLengthValueDataSize: Int = TAG_LENGTH_VALUE_MAXIMUM_VALUE_SIZE,
    ): ByteArray {
        val dataLength = minOf(tagLengthValueData.size, maximumTagLengthValueDataSize)
        val buffer = ByteBuffer.allocate(1 + COMMAND_HEADER_SIZE + dataLength).order(ByteOrder.BIG_ENDIAN)
        buffer.put(Signal.COMMAND)
        buffer.putShort(sequenceNumber.toShort())
        buffer.put(commandIdentifier)
        if (dataLength > 0) buffer.put(tagLengthValueData, 0, dataLength)
        return buffer.array()
    }

    data class Command(
        val commandIdentifier: Byte,
        val sequenceNumber: Int,
        val data: ByteArray,
    )

    fun parseCommand(payload: ByteArray): Command? {
        if (payload.size < 4 || payload[0] != Signal.COMMAND) return null
        val sequenceNumber =
            ByteBuffer
                .wrap(payload, 1, 2)
                .order(ByteOrder.BIG_ENDIAN)
                .short
                .toInt() and 0xFFFF
        val commandIdentifier = payload[3]
        val data = if (payload.size > 4) payload.copyOfRange(4, payload.size) else ByteArray(0)
        return Command(commandIdentifier, sequenceNumber, data)
    }

    fun encodeTagLengthValue(
        tag: Byte,
        value: ByteArray,
    ): ByteArray {
        val length = minOf(value.size, TAG_LENGTH_VALUE_MAXIMUM_VALUE_SIZE)
        val result = ByteArray(2 + length)
        result[0] = tag
        result[1] = length.toByte()
        System.arraycopy(value, 0, result, 2, length)
        return result
    }

    fun encodeTagLengthValueString(
        tag: Byte,
        value: String,
    ): ByteArray = encodeTagLengthValue(tag, value.toByteArray(Charsets.UTF_8))

    fun encodeTagLengthValueInteger(
        tag: Byte,
        value: Int,
    ): ByteArray {
        val buffer = ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN).putInt(value)
        return encodeTagLengthValue(tag, buffer.array())
    }

    fun encodeTagLengthValueByte(
        tag: Byte,
        value: Byte,
    ): ByteArray = encodeTagLengthValue(tag, byteArrayOf(value))

    data class TagLengthValueField(
        val tag: Byte,
        val value: ByteArray,
    )

    fun decodeTagLengthValue(data: ByteArray): List<TagLengthValueField> {
        val fields = mutableListOf<TagLengthValueField>()
        var offset = 0
        while (offset + 2 <= data.size) {
            val tag = data[offset]
            val length = data[offset + 1].toInt() and 0xFF
            offset += 2
            if (offset + length > data.size) break
            val value = data.copyOfRange(offset, offset + length)
            fields.add(TagLengthValueField(tag, value))
            offset += length
        }
        return fields
    }

    fun getTagLengthValueString(
        fields: List<TagLengthValueField>,
        tag: Byte,
    ): String? = fields.firstOrNull { it.tag == tag }?.value?.toString(Charsets.UTF_8)

    fun getTagLengthValueInteger(
        fields: List<TagLengthValueField>,
        tag: Byte,
    ): Int? {
        val field = fields.firstOrNull { it.tag == tag } ?: return null
        if (field.value.size < 4) return null
        return ByteBuffer.wrap(field.value).order(ByteOrder.BIG_ENDIAN).int
    }

    fun getTagLengthValueByte(
        fields: List<TagLengthValueField>,
        tag: Byte,
    ): Byte? {
        val field = fields.firstOrNull { it.tag == tag } ?: return null
        if (field.value.isEmpty()) return null
        return field.value[0]
    }
}
